src/decode.rs

```diff
@@ -150,19 +182,42 @@ fn decode_node<I: Iterator<Item = u8>, App: Application>(
 
     debug_assert!(!index_to_node.contains_key(&index));
     index_to_node.insert(index, node);
-    Ok(())
+    Ok(index + 1)
 }
 
 /// Return the child node at the given index from a hash map.
+///
+/// If the child is a witness that is already used and witnesses should be fresh,
+/// then a witness node is inserted at the parent index,
+/// the parent index is incremented by one
+/// and the fresh witness is returned as child.
 fn get_child_from_index<App: Application>(
     index: usize,
-    index_to_node: &HashMap<usize, Rc<CommitNode<(), App>>>,
+    parent_index: &mut usize,
+    index_to_node: &mut HashMap<usize, Rc<CommitNode<(), App>>>,
+    maybe_used_witness_indices: &mut Option<HashSet<usize>>,
 ) -> Rc<CommitNode<(), App>> {
```

Comment:
In 6076bdbfe1e6624aa1c09ed9021950fe1952694c:

Is this function needed, or could you just detect witness nodes and always return a fresh `Witness` node without bothering to keep track of which indices had been used?

I think for `CommitNode`s we just want to blindly construct fresh `Witness` nodes (which should be basically free) and for `RedeemNode`s we want to enforce the ordinary sharing logic.

I also still think this logic would be easier to follow if sharing-related stuff were inlined into decoding.

Reply:
Bookkeeping is needed because nodes reference their children via indices (as in the serialisation), but at the same time the program grows and the indices (as in the deserialised program) change. At the end, we compare the order of nodes as we deserialised them with the post-order of the program's iterator. In my original code, I tried to solve both problems with the same index map, which made it complicated. To make things clearer, I rewrote the code to keep track of "indices" (as in the serialisation) and of "new indices" (as in the deserialised program).